### ISSUE TITLE
Bug: [New Expensify] Unread message shows  as bold in the LHN but green line does not appear reported by @gadhiyamanan

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#11726]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
 1. Send a message when receiver is not logged in
 2.  log into the receiver account
 4.  Check the message

## Expected Result:
Both in the LHN and in the chat shows New message indicator

## Actual Result:
No new message indicator inside the chat, only in the LHN it shows as bold as new message indicator

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.2.12-3
**Reproducible in staging?:** y
**Reproducible in production?:** y
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 
https://user-images.githubusercontent.com/43996225/195195728-9396a493-a85d-48b3-90ba-b8dc7d876f2b.mp4
https://user-images.githubusercontent.com/43996225/195195752-26184b12-3057-40dc-bac5-e3e67d76a1ff.mov
**Expensify/Expensify Issue URL:** 
**Issue reported by:** @gadhiyamanan
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1665479496661839

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by slafortune** [comment]
Also - similar issue is reported here -https://github.com/Expensify/App/issues/10736

**Comment by marcaaron** [comment]
This is a pretty frustrating one. Gonna take it internal so we can design and fix it properly, but here's why it happens:

- We are using `componentDidUpdate()` to detect that a "new action has arrived" by looking at the last action and comparing it to the latest here:

https://github.com/Expensify/App/blob/d0f0dd5ee7b88bf31d544f758ab7cfba88c6f0c0/src/pages/home/report/ReportActionsView.js#L179-L180

- We then update the `newMarkerSequenceNumber` to `0` because we are scrolled to the bottom of the chat

That is the correct behavior when you are already looking at the chat. But doesn't really work as expected when you:

- visit a chat 
- fetch more recent actions
- component updates and processes a "new action" as if it has just arrived in realtime

Said another way, we are making a really bad assumption that a "new action appeared" when we only look at the `componentDidUpdate()` data changing. A general solution here would be to only allow the `componentDidUpdate()` logic to run when a "new" report action actually appears and not just some are loaded from the server when the component mounts. But I think in that case we're better off just moving it all out of the component entirely to avoid any confusion.

**Comment by isabelastisser** [comment]
hey @marcaaron is there anything for me to do here? I reviewed the issue and your comments, and it seems like we're taking this internally, and the payment in Upwork will only need to be made _if we deploy an associated Pull Request to production._

Unassigning myself.

**Comment by JmillsExpensify** [comment]
I'll assign myself and try to find someone to take this one.

**Comment by JmillsExpensify** [comment]
Still need to circle back and find an internal volunteer. I have OOO this week, so I need to circle back on this one when I return.

**Comment by marcaaron** [comment]
I'll grab it @JmillsExpensify

**Comment by marcaaron** [comment]
Alright so kept looking into this and came up with a solution which is basically to extract the logic out of `ReportActionsView` [here](https://github.com/Expensify/App/blob/a341945917aa39adcc9ecad8e6a6d74d6a95f2c4/src/pages/home/report/ReportActionsView.js#L177-L180) (since it's entirely unreliable to detect "new actions" this way and we've been doing it since the app's inception) and instead emit an event from `Report.js` that detects whether a new action has arrived via Pusher. 

This will bring the unread indicator behavior in line with the browser notification behavior and fixes this issue (kinda).

However, there's currently a weird side effect where this only works if we wait 10 seconds between leaving a message and checking it as the recipient (because of [this issue](https://github.com/Expensify/Expensify/issues/228194)). So the new test steps would be something like:

1. Leave message as User A and wait 10 seconds
1. Open new tab to different chat as User B
1. Navigate to chat with User A as User B
1. See that new marker has been set correctly

----

I also noticed a different issue when working through this one. 

The new marker line never shows if you visit the report that is unread in a new tab.

**Repro steps for this new issue are:**

1. Leave message as User A
1. Open new session to the chat with User A as User B
1. See that new marker is not set and the chat gets marked as read in the LHN

It's correct that the report is marked as read in the LHN, but the new marker line _should_ also be present. Not sure yet why this happens, but my working theory is that we mark the report as read when you open it, but there isn't enough time to calculate the new marker position since we can only reliably do that if the `OpenApp` command has finished.

**Comment by marcaaron** [comment]
Ok so for this other issue it does look like we are initializing the report with incorrect data. If we wait until the "report data" is done loading this is still not enough. The order of operations in the `SaveResponseInOnyx` middleware and our use of `mergeCollection()` to update the report data means that data from the API is only available some time after we set the `isLoadingReportData` flag to `false` - so it's not really dependable.

Left a comment about it [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1666827378815979).

I'm not sure how to fix that without making either a change to the middleware, `react-native-onyx` or both.

**Comment by marcaaron** [comment]
Had to make a change in `react-native-onyx` to get around the second issue I came across. 
Will add focus on testing and writing up QA for the WIP PR today.

**Comment by JmillsExpensify** [comment]
Awesome, thanks a bunch @marcaaron for keeping this moving while I was out. The main point of clarity for me is whether the linked Draft PR solves both issues [you identified here](https://github.com/Expensify/App/issues/11726#issuecomment-1292708690)?

**Comment by marcaaron** [comment]
Yes it will! Writing up test steps now and will have an update soon.

**Comment by JmillsExpensify** [comment]
Ok awesome, thanks!

**Comment by marcaaron** [comment]
In review now.

**Comment by JmillsExpensify** [comment]
Linked PR waiting on confirmed tests to merge.

**Comment by JmillsExpensify** [comment]
I'll have to get to this one tomorrow. Slipped my todos today.

**Comment by marcaaron** [comment]
I think it would be fine to add a test cases for the steps in the PR. There are four test cases and I'm not sure if any of them are covered by the existing test case.

**Comment by JmillsExpensify** [comment]
Ok cool, I'll take this part then. Anything else if the checklist you think we should fill out from a CME perspective? I think the testing scripts was all that applied?

**Comment by JmillsExpensify** [comment]
Still haven't gotten to this one, though on my list.

**Comment by marcaaron** [comment]
I think this one will need to stay open based on the conversation [here](https://github.com/Expensify/App/issues/12574#issuecomment-1314277961). Unfortunately, some changes in my PR to fix this issue will need to be reverted. 

After that, I'm honestly not too sure what possible solution to apply. There is a general conflict between: showing cached content as fast as possible and showing an accurate placement for the new line indicator. If we want it fast then it will be inaccurate (at least for some amount of time). If we want an accurate placement for the new line then we have to wait for the server to tell us where it should go. By that time, it's not clear whether the behavior would be to show the "new line" or mark the report as "read".

I thought about it for a bit and couldn't come up with a clear solution. Anyways, I've worked on this code a lot so I will let some others step up and take a stab at this problem once the revert PR is up.

**Comment by JmillsExpensify** [comment]
Thanks for jumping in, as I was wavering on the best next steps. That said, what do you think about minimally ensuring that TestRail has coverage for the QA steps in #12722?

**Comment by JmillsExpensify** [comment]
Ah sorry for crossing wires. Did you ship the RCA yet? I'll make sure to give it a thorough look.

**Comment by marcaaron** [comment]
Working on it.

**Comment by JmillsExpensify** [comment]
Still need to circle back on TestRail. I'll make time for that this week.

**Comment by JmillsExpensify** [comment]
Sorry ya'll, I keep struggling to come back to this one. Crossing this off the list before Thanksgiving kicks in.

**Comment by JmillsExpensify** [comment]
Alright, apologies for the delay here. I think we are good on this regression test, which is covered by [this script](https://expensify.testrail.io/index.php?/cases/view/1971023&group_by=cases:section_id&group_order=asc&display_deleted_cases=0&group_id=229062) in TestRail. 

1. On an HT account that has a conversation with a very long history (>100 messages)
2. Navigate to the conversation that has >100 messages and Verify the loading animation is briefly displayed before the messages are loaded.
3. Verify the chat history is displayed after the loading animation
4. Verify it happens quickly and that there is no excessive delay
5. Navigate to another conversation and back to the one with >100 messages
6. Verify the loading animation is not displayed after visiting the chat that has been previously opened
7. Close the app and reopen - In Web, close and reopen the tab
8. Navigate to a conversation that has ~25 (smaller conversation) Verify the loading animation is briefly displayed before the messages are loaded.
9. Navigate to another conversation and back to the convo with ~25 messages
10. Verify the loading animation is not displayed after visiting the chat that has been previously opened

Closing this issue. Please re-open if you think I missed something.

**Comment by gadhiyamanan** [comment]
@JmillsExpensify I think the issue is not resolved yet

**Comment by JmillsExpensify** [comment]
Are you sure about that? No one else is assigned to this issue. cc @marcaaron not sure how that happened, but can you clarify?

**Comment by gadhiyamanan** [comment]
>Are you sure about that? 

still able to reproduce, I think PR reverted due to regression

cc: @JmillsExpensify @marcaaron

**Comment by marcaaron** [comment]
Yes, @gadhiyamanan is correct. See my comment [here](https://github.com/Expensify/App/issues/11726#issuecomment-1314301137).

**Comment by JmillsExpensify** [comment]
Ah thanks, I forgot about that. So what's our next step in that case? We have a daily on this issue, though we're all not treating it that way, so minimally we should agree on it's priority and update the labels/titles accordingly.

**Comment by marcaaron** [comment]
Tried reproducing the original test steps again and could not. I think we can close for now and create a new issue if it gets reported again.

https://user-images.githubusercontent.com/32969087/206552452-d5c5d660-8d31-47ad-874f-d0904c352e6f.mp4

**Comment by gadhiyamanan** [comment]
issue is reproducible on the latest version
cc: @JmillsExpensify @marcaaron  

https://user-images.githubusercontent.com/54790231/216527107-f94c50dd-716e-49a0-bd1a-2d47815250da.mov

**Comment by JmillsExpensify** [comment]
@marcaaron I'm not sure what's going on here, and I know we still have some deprecate sequence numbers polish in play. That said, I was able to reproduce the issue all the same.

**Comment by JmillsExpensify** [comment]
Actually, given that we're working on it in the linked issue from Ben and the reporter is the same, let's keep this one closed.
